src/Models/ListModel.vala

```diff
@@ -0,0 +1,147 @@
+/*
+* Copyright (c) 2019 Alecaddd (https://alecaddd.com)
+*
+* This file is part of Akira.
+*
+* Akira is free software: you can redistribute it and/or modify
+* it under the terms of the GNU General Public License as published by
+* the Free Software Foundation, either version 3 of the License, or
+* (at your option) any later version.
+
+* Akira is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+* GNU General Public License for more details.
+
+* You should have received a copy of the GNU General Public License
+* along with Akira. If not, see <https://www.gnu.org/licenses/>.
+*
+* Authored by: Giacomo "giacomoalbe" Alberini <giacomoalbe@gmail.com>
+* Authored by: Alessandro "alecaddd" Castellani <castellani.ale@gmail.com>
+*/
+
+public class Akira.Models.ListModel : GLib.Object, GLib.ListModel {
+    private GLib.List<Akira.Models.FillsItemModel?> fills_list;
+    private GLib.List<Akira.Models.BordersItemModel?> borders_list;
+    public ListType list_type;
+    public enum ListType {
+        FILL,
+        BORDER
+    }
+
+    public ListModel (ListType type) {
+        list_type = type;
+    }
+
+    construct {
```

Comment:
I think that the idea behind this refactor is good (get the logic out of `FillListModel` and put it inside a new object) but this approach is not going to scale well. 

The ListModel object should be used elsewhere too (`LayerPanel` I'm thinking about you!), so hardcoding every possible list types (with the `ListType`) is not going to work, IMHO. 

Maybe we are not going to deal with 30 different types of list, but the code might start to become convoluted and we could introduce a lot of custom logic which is not related to the `ListModel` itself in here (like the action triggered on the removal of the last element, which does something on the BordersItemModel the ListModel is and should not be resposible for).

Moreover, we are creating (and thus instantiating) one list for every "ListType" each time we use the ListModel (in this case, twice, one for each panel - fill & border -).

My suggestion is to create 2 different base class, `ListModel` and `ItemModel` in which we declare only the very specific logic (add, remove, edit, connect, etc) and then inherit the `ItemModel` to handle the custom logic we want associated with the model. 

In this case, we could create a `BorderItemModel` and `FillItemModel` in which we put the custom logic and properties. The "if this is the last element" logic could be implemented in the BorderPanel by connecting to the "on_item_remove" event of `ListModel` (or whatever the name of the event might be), since this check is more related to the BorderPanel than to the general `ListModel`. 

I don't know if this could be the matter of a more narrowed scoped PR, given the fact that the current version of `ListModel` works out just fine, but in my opinion this would benefit the logic a lot more and in the end would provide us a new truly reusable component, with every aspect or concern of the code clearly separated and in the proper context.

Reply:
That's a good point, but I wouldn't do that here.
As you suggested, I think we should implement it in a more narrowed scoped PR.
I'll open an issue for it.